### PR TITLE
Support LLVM 16

### PR DESCRIPTION
**Pull request by HertzDevil**
Tested on x86_64-linux-gnu for some build between 16.0.0-rc3 and rc4 only. No changes were required in the compiler at all because we already switched to opaque pointers for LLVM 15.

This does not include the debug information stuff in #13174.

### CODE REVIEW

**Comment by HertzDevil** [comment]
This workaround:

https://github.com/crystal-lang/crystal/blob/3a360b1bfc966474c2e9f21594c7b4030d29ac73/.github/workflows/win.yml#L309-L313

is also unnecessary for LLVM 16 due to https://reviews.llvm.org/rG36357c967cdb9006dd7de9be0dd52068da2a4f96

**Comment by HertzDevil** [comment]
`/mnt/llvm/bin/llvm-config: error while loading shared libraries: libtinfo.so.5: cannot open shared object file: No such file or directory`

🤔

**Review by straight-shoota — APPROVED** [review]
🚀

File: .github/workflows/llvm.yml [hunk 1]
```diff
@@ -36,6 +36,10 @@ jobs:
           version: "${{ matrix.llvm_version }}"
           cached: ${{ steps.cache-llvm.outputs.cache-hit }}
 
+      - name: Install libtinfo5
+        run: apt-get install -y libtinfo5
+        if: "{{ matrix.llvm_version == '16.0.0' }}"
```
Comment (Sija) [inline .github/workflows/llvm.yml:41]: No need for interpolation.

```suggestion
        if: matrix.llvm_version == "16.0.0"
```

File: .github/workflows/llvm.yml [hunk 2]
```diff
@@ -36,6 +36,10 @@ jobs:
           version: "${{ matrix.llvm_version }}"
           cached: ${{ steps.cache-llvm.outputs.cache-hit }}
 
+      - name: Install libtinfo5
+        run: apt-get install -y libtinfo5
+        if: "${{ matrix.llvm_version == '16.0.0' }}"
```
Comment (Sija) [inline .github/workflows/llvm.yml:41]: No need for interpolation:

```suggestion
        if: matrix.llvm_version == "16.0.0"
```

File: .github/workflows/llvm.yml [hunk 3]
```diff
@@ -16,6 +16,10 @@ jobs:
       fail-fast: false
       matrix:
         llvm_version: ["13.0.0", "14.0.0", "15.0.6"]
+        include:
+          - llvm_version: "16.0.3"
+            force_version: true
```
Comment (straight-shoota) [inline .github/workflows/llvm.yml:21]: Do you happen to know why 16.0.3 is not available as a supported version that we need to force it?

**Review by straight-shoota — APPROVED** [review]

File: .github/workflows/llvm.yml (hunk 3 above)
Comment (HertzDevil) [reply .github/workflows/llvm.yml:21]: It is only unavailable from the [GitHub action that we use](https://github.com/KyleMayes/install-llvm-action)

Comment (straight-shoota) [reply .github/workflows/llvm.yml:21]: Should we submit a patch to add it then?

Comment (HertzDevil) [reply .github/workflows/llvm.yml:21]: All it does is fetch from llvm-project, which means the Windows installations do not have `llvm-config.exe`. For macOS we obtain LLVM via Nix. So not much is gained from this degree of portability

Comment (straight-shoota) [reply .github/workflows/llvm.yml:21]: I'm happy to revert to an explicit download and extraction if it improves things.

**Review by straight-shoota — APPROVED** [review]

File: .github/workflows/llvm.yml [hunk 4]
```diff
@@ -15,39 +15,41 @@ jobs:
...
       fail-fast: false
       matrix:
-        llvm_version: ["13.0.0", "14.0.0", "15.0.6"]
         include:
+          - llvm_version: "13.0.0"
+            llvm_ubuntu_version: "20.04"
+          - llvm_version: "14.0.0"
+            llvm_ubuntu_version: "18.04"
+          - llvm_version: "15.0.6"
+            llvm_ubuntu_version: "18.04"
```
[the remark this answers is not in the document]
Comment (HertzDevil) [reply .github/workflows/llvm.yml:24]: That's the version number that comes with the LLVM download, our base image is now always 22.04

**Pull request merged**